### PR TITLE
Add memcache based shared/exclusive locking

### PR DESCRIPTION
**Pull request by icewind1991**
This adds shared/exclusive locking based on the various memcache backends

To get proper atomic locking the memcache interface/implementations have been extended with atomic add (set if not exists), inc, dec and cas (compare and swap)

@PVince81 @DeepDiver1975 as discussed

### CODE REVIEW

**Comment by PVince81** [comment]
Nice work!

I get this is only the ground work that adds the new classes ? I didn't see the providers instantiated anywhere.

File: lib/private/memcache/xcache.php [hunk 1]
```diff
@@ -67,11 +69,73 @@ public function clear($prefix='') {
...
+
+	/**
+	 * Compare and set
+	 *
+	 * @param string $key
+	 * @param mixed $old
+	 * @param mixed $new
+	 * @return bool
+	 */
+	public function cas($key, $old, $new) {
```
Comment (Xenopathic) [inline lib/private/memcache/xcache.php]: Isn't this the same as CASTrait?

**Comment by Xenopathic** [comment]
Memcache bits look good, no obvious race conditions.

**Comment by DeepDiver1975** [comment]
> I get this is only the ground work that adds the new classes ?

Looks like this can easily be reviewed and merge - independent of the actual file/folder locking mechanism - which is actually a great thing

@icewind1991 ready to review then?

File: lib/private/memcache/xcache.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/memcache/xcache.php]: good catch, forgot to remove this after moving it to the trait

File: lib/private/memcache/castrait.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	/**
+	 * Compare and set
+	 *
+	 * @param string $key
+	 * @param mixed $old
+	 * @param mixed $new
+	 * @return bool
+	 */
+	public function cas($key, $old, $new) {
```
Comment (DeepDiver1975) [inline lib/private/memcache/castrait.php]: unit testing this would make sense from my understanding

Comment (icewind1991) [inline lib/private/memcache/castrait.php]: done

File: lib/public/lock/lockedexception.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP\Lock;
+
+class LockedException extends \Exception {
+}
```
Comment (PVince81) [inline lib/public/lock/lockedexception.php]: Do we want a special field for the path ?

Remember that we'll need to translate this to a Sabre Exception in some cases, and it lets us pass lock info: https://github.com/owncloud/3rdparty/blob/master/sabre/dav/lib/DAV/Exception/Locked.php#L33

On the other hand it is possible that the Sabre Exception is designed specificall for WebDAV locks, not custom locks.

Comment (DeepDiver1975) [inline lib/public/lock/lockedexception.php]: we can extend this as soon as we integrate with webdav

**Comment by icewind1991** [comment]
@owncloud-bot retest this please?

**Comment by PVince81** [comment]
I agree for merging this if it does not affect anything else and is only a pure addition. Then let's do the integration with files separately.

:+1:

**Comment by karlitschek** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
@owncloud-bot retest this please?

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by DeepDiver1975** [comment]
:+1:  :shipit:

**Pull request merged**